File: sce.m
function Y = sce(P, maxIter, alpha, weights, Y0, eta0, bConstantEta, nRepuSamp, blockSize, blockCount)
nn = size(P,1);
if ~exist('weights', 'var') || isempty(weights)
    weights = ones(nn,1);
end
if ~exist('alpha', 'var') || isempty(alpha)
    alpha = 0.5;
end
if ~exist('Y0', 'var') || isempty(Y0)
    rng(0, 'twister');
    Y0 = randn(nn,2)*1e-4;
end
if ~exist('maxIter', 'var') || isempty(maxIter)
    maxIter = 1e5;
end
if ~exist('eta0', 'var') || isempty(eta0)
    eta0 = 1;
end
if ~exist('nRepuSamp', 'var') || isempty(nRepuSamp)
    nRepuSamp = 1;
end
if ~exist('blockSize', 'var') || isempty(blockSize)
    blockSize = 128;
end
if ~exist('blockCount', 'var') || isempty(blockCount)
    blockCount = 128;
end
if ~exist('bConstantEta', 'var') || isempty(bConstantEta)
    bConstantEta = 0;
end
if bConstantEta~=0
    bConstantEta = 1;
end


[I,J,V] = find(P);
ne = length(I);

fnameP = tempname;
fid = fopen(fnameP, 'w+');
fwrite(fid, [nn ne], 'uint64');
fwrite(fid, I-1, 'uint64');
fwrite(fid, J-1, 'uint64');
fwrite(fid, V, 'double');
fclose(fid);

fnameWeights = tempname;
fid = fopen(fnameWeights, 'w+');
fwrite(fid, weights, 'double');
fclose(fid);

fnameY0 = tempname;
fid = fopen(fnameY0, 'w+');
fwrite(fid, Y0', 'float');
fclose(fid);

fnameY = tempname;

prog = 'sce';
cmd_str = sprintf('./%s 1 %s %s %s %s %d %f %d %d %d %f %d', ...
    prog, fnameP, fnameY, fnameWeights, fnameY0, maxIter, eta0, nRepuSamp, blockSize, blockCount, alpha, bConstantEta);

status = system(cmd_str);

fid = fopen(fnameY);
res = textscan(fid, '%f %f', 'CollectOutput', 1);
fclose(fid);
Y = res{1};

delete(fnameP);
delete(fnameY);
delete(fnameWeights);
delete(fnameY0);

File: sce.py
import matplotlib.pyplot as plt
import numpy as np
import tempfile
import os
import random


def display_visualization(ycoords, labels=None, marker_size=10):
    n = ycoords.shape[0]
    if labels is None:
        labels = np.ones(n)
    nc = np.max(labels) + 1
    fig = plt.figure(figsize=(8, 8), dpi=120)
    colors = plt.cm.jet(np.linspace(0, 1, nc))
    for ci in range(nc):
        ind = labels == ci
        plt.scatter(ycoords[ind, 0], ycoords[ind, 1], marker='.', s=marker_size, color=colors[ci])
    plt.axis('off')
    fig.tight_layout()
    plt.show()


def sce(sim_mat, max_iter=10**5, alpha=0.5, weights=None, ycoords0=None, eta0=1,
        constant_eta=False, n_repu_samp=1, block_size=128, block_count=128, random_seed=0):

    random.seed(random_seed)
    np.random.seed(random_seed)

    nn = sim_mat.shape[0]
    ne = sim_mat.nnz

    fsim = tempfile.NamedTemporaryFile(delete=False)
    fsim.write(np.array(nn).astype(np.uint64).tobytes())
    fsim.write(np.array(ne).astype(np.uint64).tobytes())
    fsim.write(sim_mat.row.astype(np.uint64).tobytes())
    fsim.write(sim_mat.col.astype(np.uint64).tobytes())
    fsim.write(sim_mat.data.astype(np.double).tobytes())
    fsim.close()

    if weights is None:
        weights = np.ones(nn)
    fweights = tempfile.NamedTemporaryFile(delete=False)
    fweights.write(weights.astype(np.double).tobytes())
    fweights.close()

    if ycoords0 is None:
        ycoords0 = np.random.randn(nn, 2)
    fycoords0 = tempfile.NamedTemporaryFile(delete=False)
    fycoords0.write(ycoords0.astype(np.single).tobytes())
    fycoords0.close()

    fycoords = tempfile.NamedTemporaryFile(delete=False)
    fycoords.close()

    if not constant_eta:
        constant_eta = 0
    else:
        constant_eta = 1

    prog = 'sce'
    cmd_str = "./%s 1 %s %s %s %s %d %f %d %d %d %f %d" % \
              (prog, fsim.name, fycoords.name, fweights.name, fycoords0.name,
               max_iter, eta0, n_repu_samp, block_size, block_count, alpha, constant_eta)
    os.system(cmd_str)

    ycoords = np.loadtxt(fycoords.name)

    os.remove(fsim.name)
    os.remove(fweights.name)
    os.remove(fycoords.name)
    os.remove(fycoords0.name)

    return ycoords


